### DISCUSSION TITLE
Request Enhancement: trim

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#4858] [Development] [answered]
I would like to be able to access the parts of the the trim from %@. 

So the following new string formats would be useful:

```
%[trim]
%[trim:width]
%[trim:height]
%[trim:xoffset]
%[trim:yoffset]

```

**Comment by GeeMack** [comment]
I agree this could be helpful for several situations. I have commands where I access the trim dimensions to adjust font alignments. Something like this creates a variable to use elsewhere in the command...

    . . .
    label:"O" ^
    -set page %[@] ^
    -set option:offset %[fx:(h-page.y-page.height-page.y)/2] ^
    +delete ^
    . . .

Or to adjust the vertical centering of line of text within its envelope I might use something like this...

    magick -pointsize 100 label:"THIS IS A TEST" -set page %[@] ^
       -extent %[fx:w]x%[fx:h*2]+0-%[fx:(h-page.y-page.height-page.y)/2+(h/2)] test.png

Either of these cases, and lots of other alignment operations, could be simplified with direct access to `%[trim]` dimensions.

I would want to be able to specify which image I'm accessing in the list, `u`, `v`, `s`, `u[4]`, etc., maybe like this...

    %[trim:u[3].width]
    %[trim:v.height]
    %[trim:u[-1].xoffset]
    %[trim:s.yoffset]

I'm not sure how we'd specify that if we wanted the whole set of trim dimensions, something like this?...

    %[trim:u[2]]

So generally yes, I see how accessing trim dimensions would be usable addition.

**Reply by fmw42** [reply]
You might not need the image, if you put the trim arguments in parentheses with each image and use them as -set option:trim_width "%[trim:width]". Then later you can use any set option argument in an FX: expression or combinations of them.  However, this is a bit less compact than using the image u[id] in the expression. But I do not know how easy that would be to code.

**Reply by GeeMack** [reply]
I don't know the ease or practicality of coding, but... Would it also be useful to do calculations within the brackets?...

    -set option:myvar %[trim:(u.width+u[-1].width)/2]

**Comment by fmw42** [comment]
I think the trim options would be easier to use if make into fx: options rather than string formats, if that is possible. Then one could do them on the u[id]. They would be similar to page arguments in that regard.

**Comment by urban-warrior** [comment, marked as answer]
The trim offsets are encoded in the virtual page.  For this image:

![x1](https://user-images.githubusercontent.com/12486802/154967958-6bbe69da-7932-4f7e-953d-04701a8f80fe.png)

when trimmed, returns a page offset of 150x150+40+52.  Use `%[fx:page.width]` for the equivalent of your ask, `%[trim:width]`.

**Reply by fmw42** [reply]
OK. Thanks. I overlooked that connection.
